nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -224,229 +940,150 @@ in
-      };
-    };
+    systemd = {
+      # Xen provides systemd units.
+      packages = [ cfg.package ];
 
+      mounts = [
+        {
+          description = "Mount /proc/xen files";
+          what = "xenfs";
+          where = "/proc/xen";
+          type = "xenfs";
+          unitConfig = {
+            ConditionPathExists = "/proc/xen";
+            RefuseManualStop = "true";
+          };
+        }
+      ];
 
-    systemd.services.xen-console = {
-      description = "Xen Console Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "xen-store.service" ];
-      requires = [ "xen-store.service" ];
-      preStart = ''
-        mkdir -p /var/run/xen
-        ${optionalString cfg.trace "mkdir -p /var/log/xen"}
-        grep -q control_d /proc/xen/capabilities
-        '';
-      serviceConfig = {
-        ExecStart = ''
-          ${cfg.package}/bin/xenconsoled\
-            ${optionalString ((builtins.compareVersions cfg.package.version "4.8" >= 0)) " -i"}\
-            ${optionalString cfg.trace " --log=all --log-dir=/var/log/xen"}
+      services = {
+        xenstored = {
+          wantedBy = [ "multi-user.target" ];
+          preStart = ''
+            export XENSTORED_ROOTDIR="/var/lib/xenstored"
+            rm -f "$XENSTORED_ROOTDIR"/tdb* &>/dev/null
+            mkdir -p /var/{run,log,lib}/xen
           '';
-      };
-    };
-
+        };
 
-    systemd.services.xen-qemu = {
-      description = "Xen Qemu Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "xen-console.service" ];
-      requires = [ "xen-store.service" ];
-      serviceConfig.ExecStart = ''
-        ${cfg.package-qemu}/${cfg.package-qemu.qemu-system-i386} \
-           -xen-attach -xen-domid 0 -name dom0 -M xenpv \
-           -nographic -monitor /dev/null -serial /dev/null -parallel /dev/null
-        '';
-    };
+        xen-init-dom0 = {
+          restartIfChanged = false;
+          wantedBy = [ "multi-user.target" ];
+        };
 
+        xen-qemu-dom0-disk-backend = {
+          wantedBy = [ "multi-user.target" ];
+          serviceConfig = {
+            PIDFile = cfg.qemu.pidFile;
+            ExecStart = ''
+              ${cfg.qemu.package}/${cfg.qemu.package.qemu-system-i386} \
+              -xen-domid 0 -xen-attach -name dom0 -nographic -M xenpv \
+              -daemonize -monitor /dev/null -serial /dev/null -parallel \
+              /dev/null -nodefaults -no-user-config -pidfile \
+              ${cfg.qemu.pidFile}
+            '';
+          };
+        };
 
-    systemd.services.xen-watchdog = {
-      description = "Xen Watchdog Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "xen-qemu.service" "xen-domains.service" ];
-      serviceConfig.ExecStart = "${cfg.package}/bin/xenwatchdogd 30 15";
-      serviceConfig.Type = "forking";
-      serviceConfig.RestartSec = "1";
-      serviceConfig.Restart = "on-failure";
-    };
+        xendriverdomain.wantedBy = [ "multi-user.target" ];
```

Comment:
Shouldn't this only be activated in a driver domain domU, i.e. not on a dom0?

Also not sure exactly how the provided systemd service looks like (probably correct), but I've had to add `requires = [ "proc-xen.mount" ]` previously.

Reply:
Hmm. Since it was together with the other dom0 units, I assumed it should be loaded together.

Since it needs to be available in a domU, I think the way to go about this is to add another output to the `xen` package that includes solely that unit, and remove it from the usual `out` output. We can then add a `virtualisation.xen.domU.driverDomain` option to domUs and have it enable the service.